integration_tests/test_dataset_add.py

```diff
@@ -0,0 +1,410 @@
+    assert r.exit_code != 0
+    assert 'Use --confirm-ignore-lineage from non-interactive scripts' in r.output
+
+
+def check_bad_yaml(clirunner, index):
+    prefix = write_files({'broken.yml': '"'})
+    r = clirunner(['dataset', 'add', str(prefix/'broken.yml')])
+    assert 'ERROR Failed reading documents from ' in r.output
+
+
+def test_dataset_add(dataset_add_configs, index_empty, clirunner):
+    p = dataset_add_configs
+    index = index_empty
+    r = clirunner(['dataset', 'add', p.datasets], expect_success=False)
+    assert r.exit_code != 0
+    assert 'Found no products' in r.output
+
+    clirunner(['metadata_type', 'add', p.metadata])
+    clirunner(['product', 'add', p.products])
+    clirunner(['dataset', 'add', p.datasets])
+    clirunner(['dataset', 'add', p.datasets_bad1])
+
+    ds = load_dataset_definition(p.datasets)
+    ds_bad1 = load_dataset_definition(p.datasets_bad1)
+
+    r = clirunner(['dataset', 'search'], expect_success=True)
+    assert ds.id in r.output
+    assert ds_bad1.id not in r.output
+    assert ds.sources['ab'].id in r.output
+    assert ds.sources['ac'].sources['cd'].id in r.output
+
+    r = clirunner(['dataset', 'info', '-f', 'csv', ds.id])
+    assert ds.id in r.output
+
+    r = clirunner(['dataset', 'info', '-f', 'yaml', '--show-sources', ds.id])
+    assert ds.sources['ae'].id in r.output
+
+    r = clirunner(['dataset', 'info', '-f', 'yaml', '--show-derived', ds.sources['ae'].id])
+    assert ds.id in r.output
+
+    ds_ = SimpleDocNav(gen_dataset_test_dag(1, force_tree=True))
+    assert ds_.id == ds.id
+
+    x = index.datasets.get(ds.id, include_sources=True)
+    assert str(x.sources['ab'].id) == ds.sources['ab'].id
+    assert str(x.sources['ac'].sources['cd'].id) == ds.sources['ac'].sources['cd'].id
+
+    check_skip_lineage_test(clirunner, index)
+    check_no_product_match(clirunner, index)
+    check_with_existing_lineage(clirunner, index)
+    check_inconsistent_lineage(clirunner, index)
+    check_missing_metadata_doc(clirunner)
+    check_missing_lineage(clirunner, index)
+    check_no_confirm(clirunner, p.datasets)
+    check_bad_yaml(clirunner, index)
+
+    # check --product=nosuchproduct
+    r = clirunner(['dataset', 'add', '--product', 'nosuchproduct', p.datasets],
+                  expect_success=False)
+
+    assert "ERROR Supplied product name" in r.output
+    assert r.exit_code != 0
+
+    # Check that deprecated option is accepted
+    r = clirunner(['dataset', 'add', '--auto-match', p.datasets])
+    assert 'WARNING --auto-match option is deprecated' in r.output
+
+
+def test_dataset_add_ambgious_products(dataset_add_configs, index_empty, clirunner):
+    p = dataset_add_configs
+    index = index_empty
+    mk = dataset_maker(0)
+
+    ds = SimpleDocNav(mk('A',
+                         product_type='eo',
+                         flag_a='a',
+                         flag_b='b'))
+
+    prefix = write_files({
+        'products.yml': '''
```

Comment:
[`textwrap.dedent`](https://docs.python.org/3.7/library/textwrap.html#textwrap.dedent) is a pretty way to format this sort of code.